File: java/PracticaPDOO/src/main/java/deepspace/ShieldBooster.java

/**
 *
 * @author Alejandro Villanueva Prados
 */
package deepspace;

public class ShieldBooster {
    private String name;
    private float boost;
    private int uses;
    
    ShieldBooster(String n, float b, int u) {
        name = n;
        boost = b;
        uses = u;
    }
    
    ShieldBooster(ShieldBooster otro) {
        name = otro.name;
        boost = otro.boost;
        uses = otro.uses;
    }
    
    float getBoost() {
        return boost;
    }
    
    int getUses() {
        return uses;
    }
    
    float useIt() {
        if(getUses() > 0) {
            uses--;
            return getBoost();
        }
        else
            return 1.0f;
    }
    
    /* Gracias a esta sobrecarga, obtenemos mucha información sobre el objeto, 
    sobre todo para comparar que estamos realizando la copia correctamente */
    @Override
    public String toString() {
        return "ShieldBooster[" + "@" + System.identityHashCode(this) + ", " + name + ", " + getBoost() + ", " + getUses() + "]";
    }
}

File: ruby/ShieldBooster.rb
require_relative "ShieldToUI"

module Deepspace
  class ShieldBooster
    def initialize(n, b, u)
      @name = n
      @boost = b
      @uses = u
    end

    def self.newCopy(otro)
      otro.clone
    end

    attr_reader :name, :boost, :uses

    def getBoost
      @boost
    end

    def getUses
      @uses
    end

    def useIt
      if @uses > 0
        @uses -= 1
        @boost
      else
        1.0
      end
    end
    
    def getUIversion
      ShieldToUI.new(self)
    end

    def to_s
      "Nombre: #{name} - Potenciador: #{boost} - Usos: #{uses}"
    end

    def clone
      s = ShieldBooster.new(@name, @boost, @uses)
    end
  end # class
end # module
